### PR TITLE
Simplify explosion iteration count calculation

### PR DESCRIPTION
**Pull request by kmecpp**
Simplifies the nuclear/thermobaric explosion iteration count calculation. The following graph demonstrates the new equation is functionally equivalent: https://www.desmos.com/calculator/kdshnrxxpk

### CODE REVIEW

**Comment by DarkGuardsman** [comment]
This will greatly change how the raytraces work. Can you pull example data from our [calculation app ](https://github.com/DarkGuardsman/ICBM-Mod-Calculator) to show it doesn't impact functionality.

**Comment by kmecpp** [comment]
The only thing that's changing is the **_number_** of rays that are traced, and it's only changing that number by at most 1 (usually 0) for all radii greater than 1. See the following code 

	private static float oldVersion(float radius) {
		return (int) Math.ceil(Math.PI / Math.atan(1.0D / radius));
	}

	private static int newVersion(float radius) {
		return (int) Math.ceil(Math.PI * radius);
	}

	public static void main(String[] args) {
		for (int r = 0; r < 100; r++) {
			//		for (float r = 0; r < 100; r += 0.1) {
			System.out.println((Math.round(r * 10) / 10) + ") " + oldVersion(r) + " == " + newVersion(r) + " Delta: " + (newVersion(r) - oldVersion(r)));
		}
    }

The curves become equal as the radius goes to infinity, so the larger the radius, the closer they will match.

And from the calculation app:
OLD:
image

NEW:
image

**Comment by kmecpp** [comment]
Here's a comparison where the step count differs by 1 (radius = 14)

OLD:
image

NEW:
image

**Comment by DarkGuardsman** [comment]
Ok let me do a final check on large sizes and I'll merge it when I get home from the doctor.

**Comment by DarkGuardsman** [comment]
Ya something is off, so I need to do more checks
image

**Comment by kmecpp** [comment]
Looks like there's something wrong with Equation A there. See https://docs.google.com/spreadsheets/d/1pvnw5nhAkdINahOX-w_31VqmqiL3FEc61oKtSaXRSZ4/edit?usp=sharing

**Comment by DarkGuardsman** [comment]
Ya I see what I did `*` instead of `/` and it looks good to merge. Any chance you know why the reduction works and possibly the reason the old method was used? As this is some old code and would be useful to document some of the thought process behind it.

**Comment by kmecpp** [comment]
I've had the chance to look into this more now and I've found some pretty interesting things.

Not entirely sure how the original equation was derived but I believe Calclavia probably started off by finding the angle between rays so that the distance between the rays at the perimeter of the circle was 1 unit/block (it's fairly obvious that the distance between rays can't be more than 1 block at the perimeter otherwise you'd be skipping blocks at the perimeter). This angle can easily be found to be 1/r using the arc length formula s=r\*theta. Then somehow the PI and atan get thrown in there, not really sure how because the obvious next thing do to in my mind to get the steps, would be to see how many times this angle fits into the full 360 degrees = 2\*PI radians of a circle. Thus you get 2*PI/(1/r)=2\*PI\*R which is just my new formula (yes the formula merged was actually PI\*R which is equivalent to the old formula but aparently this was always wrong and I'll get more into that later)

The new formula is actually even more intuitive than this though. Once you realize that there cannot be more than a one block separation between rays at the perimeter of the blast, this immediately gives us a lower bound of 2\*PI\*R steps, since the the circumference has a length of 2\*PI\*R blocks so if each block gets its own ray, then we'll need 2\*PI\*R blocks.

I didn't come up with that initially though. If I remember correctly, the motivation for all of this was that I was trying to figure out the meaning behind that equation so I wrote a program to brute force an answer to the problem of "_how many rays is required to hit every block in the radius_". With my brute force calculator I started with trying to find a linear relationship and that led me to find that PI\*R seemed to produce a very good approximation. That's when I realized that the original function is just an unnecessarily complex version of the same relationship. However PI\*R as purely the number of steps is actually not a good approximation and it was just a coincidence and accident I saw those match up. I thought it was PI\*R because the way I was generating my circle was by doing each quadrant in parallel. Thus, I was only doing the steps on 1/4th of the circle so the actual value for steps that I was looking at was 4\*PI\*R.

The reason why PI*R (or PI / atan(1/r)) steps works okay in the code is because the ray is incremented by 0.5 instead of 1 so this adds an effective factor of 2 to the coverage area. This is actually still not sufficient though, because you really need both a step value of 2\*PI\*R AND a ray increment of 0.5 to get a coverage area similar to 4\*PI\*R. The mod currently (and has for years) just used PI\*R and 0.5 which as shown at the bottom doesn't cover everything. [However, interestingly enough, it does use the correct step value for phi by multiplying steps by 2, just not for theta](https://github.com/BuiltBrokenModding/ICBM-Classic/blob/6ec6ab695a0fadc4a6b14017c71dee042d43bb92/src/main/java/icbm/classic/content/blast/threaded/BlastNuclear.java#L72)

Side Note: Even though Steps = 2\*PI\*R and RAY_INCREMENT=0.5 gives an extremely good approximation, even this will start to miss some blocks around R=180. The relationship might not even be linear.

Also the calculation app is inconsistent with the mod here. [It uses 4\*PI\*R by multiplying steps by 2 here](https://github.com/DarkGuardsman/ICBM-Mod-Calculator/blob/ae7b6a16dfe61167b2d231ba3c5c6f4840b4ee63/src/main/java/icbm/classic/app/test/tools/blast/PanelLargeBlast.java#L454). This makes the calculation app appear that the rays are hitting every block while in the actual mod it's not even close.

(Also somewhat related and possibly the reason why the inconsistency was introduced is that theta_n is incorrectly named and should be named phi_n in the calculation app function)

So in summary:
- The calculation app is inconsistent with the mod
- The step calculation in the mod has always been buggy. The correct value for steps is 2\*PI\*R with ray increment of 0.5 (which is used for phi but not theta. However, this will be twice as slow and considering no one has noticed, it's probably not a huge deal

STEPS = PI\*R, RAY_INCREMENT=0.5 (what is currently used by the mod and what has been used in the mod for years)
image

STEPS = 2\*PI\*R, RAY_INCREMENT=0.5 (the correct parameters. what is currently used by the calculation app)
image

Steps = 2\*PI\*R, RAY_INCREMENT=1
image

Steps = 4\*PI\*R, RAY_INCREMENT=1
image

**Comment by kmecpp** [comment]
I've had the chance to look into this more now and I've found some pretty interesting things.

Not entirely sure how the original equation was derived but I believe Calclavia probably started off by finding the angle between rays so that the distance between the rays at the perimeter of the circle was 1 unit/block (it's fairly obvious that the distance between rays can't be more than 1 block otherwise you'd be skipping blocks at the perimeter). This angle can easily be found to be 1/r using the arc length formula s=r*theta. Then somehow the PI and atan come get thrown in there, not really sure how because the obvious next step in my mind to get the steps would be to see how many times this angle fits into the full 360 degrees = 2*PI radians of a circule. Thus you get 2*PI/(1/r)=2*PI*R which is just my new formula.

The new formula is actually even more intuitive than this though. Once you realize that there cannot be more than a one block separation between rays at the perimeter of the blast, this immediately gives us a lower bound of 2*PI*R steps, since the the circumference has a length of 2*PI*R blocks so if each block gets its own ray, then we'll need 2*PI*R blocks.

I didn't come up with that initially though. I think motivation for all of this was that I was trying to figure out the meaning behind that equation so I wrote a program to brute force an answer to the problem of how many rays is required to hit every block in the radius. I started with trying to find a linear relationship and that led me to find that PI*R seemed to produce a very good approximation. That's when I realized that the original function is just an unnecessarily complex version of the same relationship. However PI*R is actually not a good approximation and it was just a coincidence and accident I saw those match up. I thought it was PI*R because the way I was generating my circule was by doing each quadrant in parallel. Thus, I was only doing the steps on 1/4th of the circule so the actual value for steps that I was looking at was 4*PI*R.

The reason why PI*R (or PI / atan(1/r)) steps works okay in the code is because the ray is incremented by 0.5 instead of 1 so this adds an effective factor of 2 to the coverage area. This is actually still not right though, because you really need both a step value of 2*PI*R AND a ray increment of 0.5 to get a coverage area similar to 4*PI*R. The mod currently (and has for years) just used PI*R and 0.5 which as shown at the bottom doesn't cover everything.

Side Note: Even though Steps = 2*PI*R and RAY_INCREMENT=0.5 gives an extremely good approximation even this will start to miss some blocks around R=180. The relationship might not even be linear.

Also the calculation app is inconsistent with the mod here. [It uses 4\*PI\*R by multiplying steps by 2 here](https://github.com/DarkGuardsman/ICBM-Mod-Calculator/blob/develop/src/main/java/icbm/classic/app/test/tools/blast/PanelLargeBlast.java#L454). This makes the calculation app appear that the rays are hitting every block while in the actual mod it's not even close.

(Also unrelated theta_n but should be named phi_n in that function)

So in summary:
- The calculation app is inconsistent with the mod
- The step calculation in the mod has always been buggy. The correct value for steps is 2*PI*R with ray increment of 0.5 (which is used for phi but not theta. However, this will be twice as slow and considering no one has noticed, it's probably not a huge deal

STEPS = PI*R, RAY_INCREMENT=0.5 (what is currently in use and has been in use for years)
image

STEPS = 2*PI*R, RAY_INCREMENT=0.5 (the correct parameters. what is currently used by the calculation app)
image

Steps = 2*PI*R, RAY_INCREMENT=1
image

Steps = 4*PI*R, RAY_INCREMENT=1
image

**Comment by DarkGuardsman** [comment]
Ya the app is not always in sync as i toy with the math a lot. You should
look at the redmatter logic as i solved the step problem by raytracing
edges. This kills off the rotations that calcs alg uses. Also calc's is
based on a modified tnt blast.
